### ISSUE TITLE
X-LoRA: disable_adapter does not disable the experts, and generate() freezes the classifier / switches the model to eval mode

### ISSUE THREAD

**Issue by amogh-nagri-11** [huggingface/peft#3635]
### System Info

- peft: `0.20.1.dev0` (main, `15c0feda`)
- transformers: 5.16.1, torch 2.13.0, Python 3.11, macOS, CPU

### Who can help?

@BenjaminBossan @githubnemo

### Reproduction

Four related bugs in X-LoRA's adapter lifecycle. All are reproduced with the tiny model used by `tests/test_xlora.py`.

```python
import os, tempfile, torch
from transformers import AutoModelForCausalLM
from peft import LoraConfig, XLoraConfig, get_peft_model

mid = "peft-internal-testing/tiny-random-OPTForCausalLM"
tmp, files = tempfile.mkdtemp(), []
for i in range(2):
    m = AutoModelForCausalLM.from_pretrained(mid)
    pm = get_peft_model(m, LoraConfig(task_type="CAUSAL_LM", target_modules=["q_proj", "v_proj"],
                                      init_lora_weights=False))
    d = os.path.join(tmp, f"a{i}"); pm.save_pretrained(d); files.append(d)

base = AutoModelForCausalLM.from_pretrained(mid); base.config.use_cache = False
cfg = XLoraConfig(task_type="CAUSAL_LM", hidden_size=base.config.hidden_size, xlora_depth=2,
                  adapters={str(i): f for i, f in enumerate(files)})
model = get_peft_model(base, cfg).eval()
ids = torch.randint(0, 100, (1, 6))

raw = AutoModelForCausalLM.from_pretrained(mid); raw.config.use_cache = False; raw.eval()
with torch.no_grad():
    ref = raw(input_ids=ids).logits
    on = model(input_ids=ids).logits
    with model.disable_adapter():
        off = model(input_ids=ids).logits
print("1) enabled vs base :", (on - ref).abs().max().item())
print("1) disabled vs base:", (off - ref).abs().max().item())   # expected 0.0

print("2) experts trainable after one forward:",
      sum(p.requires_grad for n, p in model.named_parameters() if "lora_A" in n or "lora_B" in n))

model.train()
model.generate(input_ids=ids[:, :4], max_new_tokens=4)
print("3) inner model training mode after generate:", model.base_model.lora_model.model.training)
print("4) classifier params still trainable:",
      sum(p.requires_grad for n, p in model.named_parameters() if "internal_xlora_classifier" in n))
```

Output on `main`:

```text
1) enabled vs base : 0.4138830602169037
1) disabled vs base: 0.5074496269226074
2) experts trainable after one forward: 40
3) inner model training mode after generate: False
4) classifier params still trainable: 0
```

**1. `disable_adapter()` does not disable X-LoRA.** `XLoraLinearLayer.forward` (and the Embedding/Conv2d variants) replace the wrapped LoRA layer's `forward` and never look at the disabled state. `XLoraModel.disable_adapter_layers()` only sets `self.disabled = True`, which stops `_enable_peft_forward_hooks` from injecting scalings — so inside the context every expert is applied with `scaling_weight = 1` instead of not being applied. Disabling the adapter therefore moves the output *further* from the base model. `tests/test_xlora.py::test_disable_adapter` only asserts that the outputs differ, so it passes.

**2. Frozen experts are silently unfrozen by a forward pass.** The scalings pass calls `self.lora_model.disable_adapter_layers()` and then `enable_adapter_layers()` in a `finally`; the latter goes through `set_adapter`, which sets `requires_grad=True` on all experts. `_maybe_freeze_all_adapters()` is only called after `generate()`, so a plain `forward` leaves all experts trainable even with the default `use_trainable_adapters=False`.

**3. `generate()` permanently switches the model into eval mode.** `_maybe_freeze_all_adapters()` calls `self.eval()`, and `generate()` calls it on every invocation. Generating once during training (evaluation, RL-style rollouts) disables dropout for the rest of the run. `PeftModel.training` still reports `True`, so this is invisible.

**4. `generate()` freezes the X-LoRA classifier.** `_maybe_freeze_all_adapters()` matches parameters with `"lora_" in name` while iterating over the whole `XLoraModel`. The classifier parameters are named `internal_xlora_classifier.layers.*`, and `"internal_xlora_classifier"` contains `"lora_"`. At construction time the classifier does not exist yet, so it survives; on the call after `generate()` it is frozen. Since the classifier is the only trainable part of X-LoRA when the experts are frozen, training after a `generate()` call is a silent no-op.

Additionally, the Hub-loading branch of `PeftModel.from_pretrained` builds the expert paths as `os.path.join(model_id, model_id)` (`peft_model.py:575`), which is not a valid repo id — the subfolder is passed separately. That path is currently unreachable anyway because of the unpacking bug on `xlora/model.py:270`, which is already fixed in #3421.

### Expected behavior

- `disable_adapter()` reproduces the base model output.
- Experts stay frozen unless `use_trainable_adapters=True`, regardless of how many forward passes were run.
- `generate()` does not change the model's training mode.
- The X-LoRA classifier stays trainable.

### Would a PR be welcome?

I have a fix ready with five regression tests that fail on `main` and pass with the change (`pytest tests/test_xlora.py`: 21 passed; on `main` the five new tests fail). Per the contribution guidelines I'll wait for a maintainer go-ahead before opening it.

### AI assistance disclosure

AI assistance was used to find and prepare this. I have reviewed every line and run the reproductions and the test suite myself.
